File: src/env.ts
import { isRecord } from './helper.js';
import type { CustomArgs, CustomArgValue } from './types.js';

export const envKey = 'PLAYWRIGHT_ARGS_JSON';

export function encodeEnvArgs(customArgs: CustomArgs): Record<typeof envKey, string> {
  return {
    [envKey]: JSON.stringify(customArgs),
  };
}

export function decodeEnvArgs(env: NodeJS.ProcessEnv = process.env): CustomArgs {
  const value = env[envKey];
  if (!value) return {};

  let parsed: unknown;
  try {
    parsed = JSON.parse(value);
  } catch (error) {
    throw new Error(`${envKey} contains invalid JSON`, { cause: error });
  }

  if (!isCustomArgs(parsed)) throw new Error(`${envKey} contains invalid custom arguments`);

  return parsed;
}

function isCustomArgs(value: unknown): value is CustomArgs {
  return (
    isRecord(value) && Object.entries(value).every(([name, argValue]) => name !== '' && isCustomArgValue(argValue))
  );
}

function isCustomArgValue(value: unknown): value is CustomArgValue {
  if (typeof value === 'string' || typeof value === 'boolean') return true;

  return Array.isArray(value) && value.every((item) => typeof item === 'string');
}

File: src/env.rs
//! Port of `env.ts`.

use crate::parse_cli::CustomArgs;
use std::collections::HashMap;

pub const ENV_KEY: &str = "PLAYWRIGHT_ARGS_JSON";

pub fn encode_env_args(custom_args: &CustomArgs) -> HashMap<String, String> {
    let json = serde_json::to_string(custom_args).expect("custom args are always serializable");
    HashMap::from([(ENV_KEY.to_string(), json)])
}

/// Decodes from the real process environment (the `decodeEnvArgs()` default).
pub fn decode_env_args() -> Result<CustomArgs, String> {
    decode_env_value(std::env::var(ENV_KEY).ok().as_deref())
}

pub fn decode_env_value(value: Option<&str>) -> Result<CustomArgs, String> {
    let Some(value) = value.filter(|value| !value.is_empty()) else {
        return Ok(CustomArgs::new());
    };

    let parsed: serde_json::Value =
        serde_json::from_str(value).map_err(|_| format!("{ENV_KEY} contains invalid JSON"))?;

    let invalid = || format!("{ENV_KEY} contains invalid custom arguments");
    let args: CustomArgs = serde_json::from_value(parsed).map_err(|_| invalid())?;
    if args.keys().any(|name| name.is_empty()) {
        return Err(invalid());
    }

    Ok(args)
}
